src/Init/Data/Fin/Lemmas.lean

```diff
@@ -102,7 +102,13 @@ theorem dite_val {n : Nat} {c : Prop} [Decidable c] {x y : Fin n} :
     (if c then x else y).val = if c then x.val else y.val := by
   by_cases c <;> simp [*]
 
-instance (n : Nat) [NeZero n] : NatCast (Fin n) where
+/--
+This is not an instance, because it causes `x < n` (for `x : Fin k`, `n : Nat`)
+to be elaborated as `x < ↑n` rather than `↑x < n`, which is undesirable.
+Note however that in Mathlib this happens anyway!
```

Comment:
How about
```suggestion
This is not an instance because the `binop%` elaborator assumes that there are no non-trivial coercion loops,
but this introduces a coercion from `Nat` to `Fin n` and back.
Non-trivial loops lead to undesirable and counterintuitive elaboration behavior.
For example, for `x : Fin k` and `n : Nat`, it causes `x < n` to be elaborated as `x < ↑n` rather than `↑x < n`, silently introducing wraparound arithmetic.

Note: as of 2025-06-03, Mathlib has such a coercion for `Fin n` anyway!
```

Reply:
Thanks, that's a much more helpful explanation. I've included it.